### PR TITLE
Add keyboard controls for SiteSelector

### LINKED ISSUE #8: Site Picker: searching should support arrow keys + enter
The flow would be something like:

1. Select switch site
2. it should focus in the search field
3. type "my"
4. list is immediately filtered down to sites containing the string "my", with the top site in the list being auto-selected and visibly highlighted (same as the hover effect we currently have)
5. I actually want the second site in the list, so I can just hit the down arrow on my keyboard, which would move the highlight/active selection down to the second site in the list (my cursor is still in the search field though, so I could keep typing if I want)
6. I can now hit enter to select the second site in the list, since that's where my highlight/selection is.

It would work very similar as github @mentions

### LINKED ISSUE #6024: Masterbar: Auto-focus on site picker search field after clicking on new post icon
#### Steps to reproduce
1. Starting at URL: https://wordpress.com/
2. Click on new post icon in the masterbar
3. When the site picker opens, start typing or use the arrow keys to try to select a site

#### What I expected

I expect the search field in the site picker to get auto-focus, so I can immediately start searching for the right blog where I want to create a new post.

Every time I start a new post in Calypso, I run into this issue — I just expect to be able to start typing immediately after opening the site picker. It's always disappointing when I realize nothing I typed had any effect and I have to go back, click on the search field, and type again.

It also seems like a clear win to auto-focus somewhere in the site picker, since the user is expected to continue the flow by selecting a site to open the editor.

#### What happened instead

When the site picker opens, neither the search field nor the site list get focus, so typing (to search) or using the arrow keys (to browse the list) don't work. Hitting enter closes the site picker, but I need to manually click on the search field to focus on it and search.

Using the arrow keys at this point, with the expectation that you can use them to select a site in the picker, is especially jarring because it starts scrolling the Reader underneath the site picker instead.

#### Browser / OS version

Chrome 50.0.2661.102  and Firefox 47.0 / Mac OSX 10.11.5

#### Screenshot / Video

screen shot 2016-06-14 at 4 48 38 pm

#### Context / Source

#dogfooding

### PR DESCRIPTION
**Pull request by marekhrabe**
This PR adds keyboard support to every place where you can select sites. That means <kbd>Up</kbd> & <kbd>Down</kbd> arrows to move through sites in list, <kbd>Enter</kbd> to select and also <kbd>Esc</kbd> to close without changing a site.

These new keyboard features go well together with searching functionality to narrow results and that's why keystrokes are being listened directly on the search input.

Here is a quick preview of the implementation:

sites-keyboard3

# Testing Instructions

#### Places to test

- My Sites -> Switch Site (this is called `SitePicker`)
- New Post (`SitesPopover`)
- Help -> Contact us (`SitesDropdown`)
- Go to [/sites](http://calypso.localhost:3000/sites) (this is pure `SiteSelector`)

#### Each place should follow these rules

1. Mouse interactions have not changed - you can hover a site (it turns blue) and clicking a site still does the expected action
2. Search field is visible for users with >6 sites and is auto-focused. For users with <6 sites, the search should appear once they start typing some text
3. No site is initially highlighted when you open the selector.
4. Selected site has grey background
5. Pressing <kbd>Up</kbd> or <kbd>Down</kbd> changes highlighted site (and scrolls if needed)
6. With the last site in the list highlighted, pressing <kbd>Down</kbd> will select the first site and vice versa.
7. Pressing <kbd>Enter</kbd> triggers exactly the same action as clicking the highlighted site
8. Pressing <kbd>Esc</kbd> just closes the selector without any side-effects like changing a site or redirecting user
9. When you highlight some site using mouse and then press <kbd>Up</kbd> or <kbd>Down</kbd>, the highlight should move from last mouse-highlighted site (eg. not from the start of the list)
10. When you use mouse to highlight sites and then leave site selector with your cursor, the highlight should disappear and when you use arrow keys now, they should move from the beginning of list

#### Naming conventions

This is what "highlighted" site looks like. Site can be highlighted by navigating with arrow keys or by hovering on site with the mouse cursor.

screen shot 2016-06-17 at 17 03 13

This is what "selected" site looks like:

screen shot 2016-06-17 at 17 04 55

#### Altering between keyboard and mouse

I modeled those interactions after native system menus. 

select

#### Test live!

Test live: [https://calypso.live/posts?branch=add/site-selector-keyboard-accessibility](https://calypso.live/posts?branch=add/site-selector-keyboard-accessibility)

Closes #8
Closes #6024

### CODE REVIEW

**Comment by mtias** [comment]
Nice work @marekhrabe! It seems to be working rather well. One thing I noticed is that if I press the down arrow in the sidebar picker to move the highlighted site, and then wait, it eventually resets to the All Sites item being highlighted.

File: client/components/sites-dropdown/style.scss [hunk 1]
```diff
@@ -100,24 +101,6 @@
 			}
 		}
 	}
-
-	// Highlight selected site
-	&.is-selected,
```
Comment (mtias) [inline client/components/sites-dropdown/style.scss]: The behaviour of the sites dropdown is a bit awkward now, it should highlight the currently selected site when you open the dropdown, not the first in the list.

**Comment by mtias** [comment]
> it eventually resets to the All Sites item being highlighted

This seems related to the fact when you click "switch site" the highlighted site doesn't match the current site, and it should.

**Comment by mtias** [comment]
> the search field is now always visible (previously it showed up only for user that has more than 6 sites)

I don't think this is a good change. What was the reasoning for it? The picker works much better without search when you have only a few sites. Search should be an enhancement to navigation only when it makes sense.

**Comment by marekhrabe** [comment]
> One thing I noticed is that if I press the down arrow in the sidebar picker to move the highlighted site, and then wait, it eventually resets to the All Sites item being highlighted.

I suspect one of my latest changes to cause that (bc95cda1f8cd87c2414a359ebde40de805044e3f). I was able to reproduce and I'll push a fix for that soon. 

> The picker works much better without search when you have only a few sites. Search should be an enhancement to navigation only when it makes sense.

I think this might be truer for the previous version than this one. Previously, the search was something additional and optional that you could use and then you had to come back to your mouse to confirm your selection. This time, I tried to make the search a core feature of new selectors. Even if you have 4 sites, I feel like having the ability to type, for example, just the first letter and hitting <kbd>Enter</kbd> is faster than arrow navigation through the list. At least that's how I use this kind of components. My interactions with Sublime's/Atom's Go to anything were my inspiration.

> The behaviour of the sites dropdown is a bit awkward now, it should highlight the currently selected site when you open the dropdown, not the first in the list.

(Btw: that CSS has nothing to do with that. It was some legacy code that was not used, as selected sites were completely hidden in Dropdowns.)

The reasoning was that if you start with the selected site, you might run easily into edge cases: 

- selected site is always hidden in SitesDropdown (what to select now?)
- when you visit `/sites` route, there is no selected site at the moment (what to select now?)

I thought that the most reasonable way (and working in all cases), is to select the first one. When there are more solutions, I aim for predictability and consistency in all scenarios and this was the case.

As a little technical bonus, it allows us from having a double render of SiteSelector (see 329b83fa41dbf5dcacfe8a8a57f987da8025b71b), where we first need to render sites to know their order and after they are rendered, determine which one to highlight, save that to the state and rerender. Anyway, there is probably a way around that, but this is the most simple solution and they usually tend to work the best.

The only component that is really affected by this change is probably the picker (left sidebar "Switch Site"). It behaves a little different than the previous implementation, but I don't think it is a bad thing. You still do? I think that users mainly use that component to change the current site, not to see what is selected.

Thanks a lot for finding time to review this, @mtias. I would love to continue the discussion on points that I explained here if they change your point of view in some way.

**Comment by alisterscott** [comment]
I've tested this and it seems to work well 👍 
I personally don't mind the search field for < 4 sites as it's consistent and works well with the arrow keys.
Mobile regression test was fine also - I made sure that the menu page is still shown on selection.
Visiting `/sites` was broken for me, but I checked https://wordpress.com as it's broken there also so not this PR. I raised this as a separate bug: #5983

**Comment by alisterscott** [comment]
I did notice the arrow down behaves differently for non-webkit browsers (Firefox, IE11, Edge) in that the whole panel moves with the arrow, whereas Chrome moves individual sites one at a time.

For example in Firefox:

firefox scroll

is different to Chrome:

chrome scroll

**Comment by marekhrabe** [comment]
> …whole panel moves with the arrow…

Which one do you prefer, @alisterscott? The Chrome one feels better to me. It works like that probably because Webkit implements (non-standard, but commonly used) `scrollIntoViewIfNeeded` and others have just pure `scrollIntoView` (which always scrolls). We already have some kind of shim in our codebase. I'll take a second look if I can use it in this case too.

**Comment by marekhrabe** [comment]
> One thing I noticed is that if I press the down arrow in the sidebar picker to move the highlighted site, and then wait, it eventually resets to the All Sites item being highlighted.

The bug @mtias mentions should be fixed now. It now resets to the first available item only once you open the sidebar. Not in the middle of interactions. We can now discuss further if we want to change the behavior.

**Comment by folletto** [comment]
> I think this might be truer for the previous version than this one. Previously, the search was something additional and optional that you could use and then you had to come back to your mouse to confirm your selection. This time, I tried to make the search a core feature of new selectors. Even if you have 4 sites, I feel like having the ability to type, for example, just the first letter and hitting <kbd>Enter</kbd> is faster than arrow navigation through the list. At least that's how I use this kind of components. My interactions with Sublime's/Atom's Go to anything were my inspiration.

Note that the fact that previously you had to use the mouse was an incomplete implementation of the design, so it was already meant to be disappearing when less than 6 sites AND still be triggered with enter.

"Faster" here is a relative thing. We're talking of very very few keystrokes, and it's again an edge optimization that is... very edge: less than 6 sites and power user. Even there, just arrow down a few times and you're done. We're talking of a switch time still likely around the second or two.

Given the marginal gain, I wouldn't add the extra UI for users with less than 6 sites. The search examples you make (Sublime / Atom) are geared for people with terminal experience, text-heavy and in an interface primarily textual. UIs are designed around users.

So, in absence of any heavier weighting that makes tip the decision for search, I'd keep the optimization we have for a less cluttered UI, even if it costs a few milliseconds to an expert user.

> The reasoning was that if you start with the selected site, you might run easily into edge cases:
> 1. selected site is always hidden in SitesDropdown (what to select now?)
> 2. when you visit /sites route, there is no selected site at the moment (what to select now?)
> I thought that the most reasonable way (and working in all cases), is to select the first one. When there are more solutions, I aim for predictability and consistency in all scenarios and this was the case.

The two edge cases aren't really issues. If the active site isn't in the list, we don't highlight anything. It's meant to work that way (plus we have [an iteration 2](https://github.com/Automattic/calypso-pre-oss/issues/5353#issuecomment-156413765) of the `SitesDropdown` design that actually shows the selected item, centered).

The reason we highlight the current site is acting as a reference point: "this is where you are". It's a relevant marker, and has a specific function. 

Also: there are two highlight statuses already to cover the cases:

1 — Highlight as selected — the dark background:

screen shot 2016-06-13 at 16 31 34

2 — Highlight as focus — the light background

screen shot 2016-06-13 at 16 31 28

Keyboard navigation is a _focus_ kind of highlight, so shouldn't touch or change the _selected_ state.

**Comment by marekhrabe** [comment]
> Also: there are two highlight statuses already to cover the cases:

When you suggest I change the keyboard highlight color to the light colored version (the same as a mouse hover), don’t you think that might look confusing when combined with those mouse hovers? In one moment, you can have “selected site”, highlighted keyboard focus, and mouse hover in one component. I don’t say that it’s a common situation, but it might look funny and it might not be apparent what would be confirmed by pressing enter. What do you think?

**Comment by folletto** [comment]
> When you suggest I change the keyboard highlight color to the light colored version (the same as a mouse hover), don’t you think that might look confusing when combined with those mouse hovers? In one moment, you can have “selected site”, highlighted keyboard focus, and mouse hover in one component. I don’t say that it’s a common situation, but it might look funny and it might not be apparent what would be confirmed by pressing enter. What do you think?

Someone moving with the keyboard and then using the mouse, and while hovering pressing enter seems super edge case. It could be addressed just by being stateful if the interaction is triggered by mouse (once you hover, the keyboard focus is lost) or keyboard (once you use the keyboard, the mouse hover gets overridden). Doable, but imho not worth the effort given the likelyhood of the issue. I'd just go for the double highlight for now (which I think it's going to be the outcome if we don't do anything special).

**Comment by folletto** [comment]
In terms of styling, I'd note that the sidebar currently has one bug: _the focus/hover states doesn't override the selected state_.

The popover instead works as expected:

cap-

So I'd:

1. Fix the hover/focus bug.
2. Use blue as hover/focus effect instead of the pale gray background (as the popover).

**Comment by marekhrabe** [comment]
> 1. Fix the hover/focus bug.
> 2. Use blue as hover/focus effect instead of the pale gray background (as the popover).

That sounds reasonable, on it :)

**Comment by marekhrabe** [comment]
I created a PR that could eventually merge into this: #5992. I'm not really sure it is a change for better, so I'd like you to test it and hear your opinions.

**Comment by alisterscott** [comment]
> Edit: I implemented cross-browser version that is now consistent across Chrome, Firefox & Safari that I tested.

This is great. I much prefer this to the other scrolling. I tested in Firefox, IE11, Edge and Chrome and LGTM

**Comment by folletto** [comment]
> I created a PR that could eventually merge into this PR: #5992. I'm not really sure it is a change for better, so I'd like you to test it and hear your opinions.

+1 for me.

With that one done and merged, I think we still have these two pending and we're good:

1. Hide initial highlight
2. Hide search for <6 sites

**Comment by mtias** [comment]
@marekhrabe Thanks for all the work here!

I agree with @folletto that hiding search for <6 is important. We can make it still work if you start typing (search can appear once you start typing, for instance), but it seems just the arrow navigations would be plenty enough for a few sites. It's important we retain the evolution of the UI depending on how many sites you have. We also apply it in other selectors like posts and categories. Search should be there when it makes sense, not always.

Regarding the highlight, when you open the site picker it should always show the currently selected site if the url has a site fragment. That's a core navigational piece of Calypso that we shouldn't sacrifice.

At the very least we shouldn't conflate all of these potential UX changes with the ability to use keyboard nav on the site selectors, which is very valuable.

**Comment by marekhrabe** [comment]
> This is great. I much prefer this to the other scrolling. I tested in Firefox, IE11, Edge and Chrome and LGTM

@alisterscott I think the scrolling battle is unfortunately not over yet. I just noticed that the new scrolling broke these two:
 
- Help -> Contact us (SitesDropdown)
- Go to /sites (this is pure SiteSelector)

I will investigate further and hopefully fix it.

**Comment by marekhrabe** [comment]
TLDR: After some research, trials and errors, I would still like to implement a search even for users with <7 sites. It would be hidden by CSS, but in a way it can be auto-focused when opening the selector — just like for users with many sites. That would allow me to reliably listen to arrow keys + enter and as a side-effect, for those users with <7 sites, I can really easily show them the search field once they start typing something. Do you like that idea?

-----

I'm in a funny situation right now with hiding the search. Currently, I listen to arrow keys and enter directly on the search input, so I need to move listeners somewhere else. 

**First**, I tried to move it to parent div like this:

screen shot 2016-06-15 at 23 41 19

It does work, but only as long as some child is a focused `<input>`, so that doesn't solve the case where we want to remove the search.

As a **second solution**, I thought about listening for keys directly on the window object, but that might not work well because some instances of `SiteSelector` are mounted for the whole Calypso session and they are not aware whether they are opened or closed (for example `SitePicker` knows that state, but its child `SiteSelector` does not). Also, we would lose some [keyboard event normalization](https://github.com/facebook/react/blob/master/src/renderers/dom/client/syntheticEvents/SyntheticKeyboardEvent.js) that React does for us.

My **third attempt** was to investigate our `keyboard-shortcuts` and add arrow keys and enter as new entries. Unfortunately, that implementation prevents keyboard shortcuts from working when they bubble from `<input>` elements. Thus, this solution would only work with search hidden, but not with search present. 

As a **fourth solution**, I came back here and noticed @mtias mentioned this:

> We can make it still work if you start typing (search can appear once you start typing, for instance)…

and I think that it might really be the easiest solution and actually the most useful. We could hide the search field only visually, but it would still be in the DOM and (secretly) autofocused and it would be listening to arrow keys that we can handle and also for any other keys, just like a normal `<input>`. Typing something would cause the search field to appear. I think it might be the best of all worlds while being the most straight-forward to implement. What do you think?

**Comment by folletto** [comment]
> > We can make it still work if you start typing (search can appear once you start typing, for instance)…

> and I think that it might really be the easiest solution and actually the most useful.

Sounds like a good idea. Let's try it! 👍

**Comment by marekhrabe** [comment]
I'm pretty happy with the functionality right now. I've updated testing instructions in the description of this PR to reflect some design changes that we made here.

There is only one issue left that I'm aware of and that is the auto-scroll on /sites - see more https://github.com/Automattic/wp-calypso/pull/5808#issuecomment-226316430. I haven't found a good solution yet (not really a simple problem), but I don't think it's blocking this whole PR. We can eventually track it separately.

**Comment by folletto** [comment]
Looks good to me 👍

**Comment by marekhrabe** [comment]
Looks good to me too. Anyone up for a code review? 👼

**Comment by mtias** [comment]
Looking good, though I found two issues. The first one I think we should solve. The last one could be a compromise.

1) It'd be nice if we could improve the interaction between keyboard nav and mouse hovering. Right now we show two highlights:

image

We should make it so that if you are hovering, and use keyboard, we disable hovering highlight until mouse moves. Once mouse moves we disable keyboard highlight until next time user presses a key. It may be as simple as setting a class of `is-keyboard-navigation-engaged` when pressing keys or something, and remove it on next mouseenter event. What do you think @marekhrabe?

2) One problem with search input being the key event listener is that if I close search on the sidebar I can't use keyboard nav anymore — using the arrows moves the page scroll position instead.

**Comment by marekhrabe** [comment]
I would love to finish this task from my trial, seems like we are close to having this ready. I'll work on this once I'm more acquainted with my team Delta.

@mtias:

1) I think it may be most effective to solve this by merging hovers and keyboard focus into one thing. We will stop relying on CSS to create the :hover effect and we will use the same component state property to manage keyboard focus and mouse hover together. This is how OS does it for example with options of `<select>` or actually any system menu. 

select

This solution is kinda desktop-specific, as it will rely on mouse events (like mouseenter), but I think it is just fine, as we don't expect to have hover states on mobile. We have just tap and that confirms the selection right away. What do you think about this proposal? 

Regarding the second option, I think we may have a global handler that will respond to keys when our input is not focused. However, I would personally consider this as an improvement, not a requirement at this moment.

**Comment by folletto** [comment]
> I would love to finish this task from my trial, seems like we are close to having this ready.

Thank you!!! 💯 


> I think it may be most effective to solve this by merging hovers and keyboard focus into one thing. 

If we can get the behaviour of "whichever last controlling method is used defines the highlight" I think it can work (from a design perspective). :)

**Comment by lancewillett** [comment]
@marekhrabe  Would love to see this improvement merge in — what steps are left to finish it?

**Comment by marekhrabe** [comment]
@lancewillett I'm working on fixing the last problem - when you use both keyboard and mouse to control the component, it currently behaves in a weird way (last few comments here describe it). I'm aiming to get it done today

File: client/components/site-selector/index.jsx [hunk 2]
```diff
@@ -62,50 +64,160 @@ const SiteSelector = React.createClass( {
...
+			search: terms,
+			highlightedIndex: ( terms ? 0 : -1 ),
+			showSearch: ( terms ? true : this.state.showSearch ),
+			isKeyboardEngaged: false,
+		} );
+	},
+
+	componentDidUpdate( prevProps, prevState ) {
+		if ( this.state.isKeyboardEngaged && prevState.highlightedIndex !== this.state.highlightedIndex ) {
+			this.scrollToHightlightedSite();
```
Comment (ehg) [inline client/components/site-selector/index.jsx]: s/Hight/High? :)

File: client/components/site-selector/index.jsx [hunk 3]
```diff
@@ -62,50 +64,160 @@ const SiteSelector = React.createClass( {
...
+				} );
+			} else {
+				selectorElement.scrollTop = 0;
+			}
+		}
 	},
 
-	onSiteSelect( siteSlug, event ) {
-		this.closeSelector();
+	onKeyDown( event ) {
```
Comment (ehg) [inline client/components/site-selector/index.jsx]: This is a looong function to read. I wonder if splitting it up would ease the readability a bit.

File: client/components/site-selector/index.jsx [hunk 4]
```diff
@@ -165,122 +300,97 @@ const SiteSelector = React.createClass( {
...
 					ref="siteSearch"
 					onSearch={ this.onSearch }
 					autoFocus={ this.props.autoFocus }
 					disabled={ ! this.props.sites.initialized }
 					onSearchClose={ this.props.onClose }
+					onKeyDown={ this.onKeyDown }
 				/>
 				<div className="site-selector__sites" ref="selector">
 					{ this.renderAllSites() }
+					{ this.renderRecentSites() }
```
Comment (ehg) [inline client/components/site-selector/index.jsx]: Since we now have a bunch of `render*` methods, is it time to start splitting these up into components & separate files?

Comment (marekhrabe) [inline client/components/site-selector/index.jsx]: I'm sure we can move out `renderNewSiteButton` and `hidden-sites-message`, but `renderAllSites` `renderRecentSites` & `renderSites` should stay here, as the are sharing a lot of state and they also work with off-state `this.visibleSites` that we use for keyboard accessibility. It would be difficult to make this all sync across more components.

File: client/components/site-selector/index.jsx (hunk 3 above)
Comment (marekhrabe) [inline client/components/site-selector/index.jsx]: I extracted some logic to `computeHighlightedSite`. Looks better now for me, what do you think?

**Comment by marekhrabe** [comment]
I'm pretty happy with the functionality now. You can use both keyboard and mouse for interactions. Please review code & also the feeling of keyboard interactions from UX point of view. Specs and testing instructions are in the description of this PR.

**Comment by folletto** [comment]
Checked on Chrome, Safari, Firefox: 👍  

Works well! :)

**Comment by beaulebens** [comment]
This is glorious. I'm so happy this is close to release. I think it's good as it is, and I have a few suggestions to make it even better (IMHO), but I don't think they should necessarily stop us from merging what you've got (other than maybe the first one, which would change the behavior):

- as i type something, i would expect the top/first result to always be highlighted. in my case, I type "dent" to filter things, and I see 6 sites listed. one of them is "selected" (current site), but none is highlighted by default. I would expect the top one to be highlighted, and then I could hit the down arrow to pick a different one, or hit enter immediately to select the current one. this is particularly relevant when there's only one result, and i'd expect to be able to immediately hit enter to select it, vs requiring the "additional step" of hitting the down arrow to select it first.
- i seem to get a huge delay (up to a few seconds) when i first start typing, before the letters appear/things start filtering. my machine is under pretty heavy load right now, and i do have a long site list, but it makes things feel really sluggish. 
- if you've already searched for something, and switched to that site, then the next time you click to switch site, the previous search string is still in the box. in this case, I think we should highlight it so that if I start typing, it clears what's there, and I can type something new. If I want to append, then I can just hit the right arrow and start typing.
- some sort of shortcut (s?) to jump into the site picker in the first place would be awesome, so that you could switch site completely via keyboard.
- if we're going all-in with shortcuts, something like 'p' for post, or 'c' for compose might make sense to trigger the new-post button in the masterbar as well

Very exciting!

**Comment by mtias** [comment]
> i seem to get a huge delay (up to a few seconds) when i first start typing, before the letters appear/things start filtering. my machine is under pretty heavy load right now, and i do have a long site list, but it makes things feel really sluggish.

Do you notice the same on production? It'd be great to know if there is any regressions on going on.

**Comment by beaulebens** [comment]
After a little more testing I don't think it's a regression, it just seems to be part of our type/search/filter system. I haven't been using it as much recently, so hadn't noticed that it was so slow. Probably something we should look at separately.

**Comment by marekhrabe** [comment]
> as i type something, i would expect the top/first result to always be highlighted.

@beaulebens, I implemented this - It needs to be there, for sure! :) I'll test it and review what could break it.

> if you've already searched for something, and switched to that site…

Probably regression after rebasing few months of commits. I'll fix this

Thanks for testing, keep it coming :)

**Comment by mtias** [comment]
@marekhrabe I think it wasn't you, I think the regression is in master.

**Comment by JasonSnow-Mad** [comment]
@marekhrabe - I just tested in IE11 and do not see a search option:

ie11_no-search

And in Edge the search appears, but arrows do not seem to function:

edge_switch-site

Worked in FF and Chrome. Let me know if other testing would be helpful!

**Comment by lsinger** [comment]
> as i type something, i would expect the top/first result to always be highlighted.

Can confirm that this works for me (on Chrome). 

> if you've already searched for something, and switched to that site, then the next time you click to switch site, the previous search string is still in the box.

I see that as well. 

> i seem to get a huge delay (up to a few seconds) when i first start typing, before the letters appear/things start filtering. 

I seem to get that only sometimes. Feels like a "we're re-rendering too often" issue, maybe. 

> I just tested in IE11 and do not see a search option

Using the IE11 VM that Microsoft provides for testing, I wasn't able to reproduce this. Things work just as well as under macOS Chrome. @JasonSnow-Mad can you still reproduce? Maybe something was changed in the meantime? 

All in all, I don't see anything that should keep us from merging here — except the potential issue with IE11 / Edge support.

**Comment by lsinger** [comment]
> And in Edge the search appears, but arrows do not seem to function

Just finished downloading the Edge VM. The arrow-based navigation works fine there for me.

**Comment by mtias** [comment]
> I seem to get that only sometimes. Feels like a "we're re-rendering too often" issue, maybe.

There was a fix that went into master for this. We should rebase this one and see if it's fixed here as well.

**Comment by marekhrabe** [comment]
@mtias I haven't done any real benchmarks, but search feels much faster to me after rebasing. You can try: https://calypso.live/posts?branch=add/site-selector-keyboard-accessibility

@JasonSnow-Mad Can you please test search and arrow keys again? I haven't addressed any issue that could potentially fix the problem, but I just wasn't able to reproduce the problem (I used Browserstack to get IE & Edge). Looks like @lsinger had same trouble to reproduce as well. 

If we confirm with more people that IE & Edge are no longer a problem, I think we might be ready to go. 

There is one regression now in `SitePicker `:

> if you've already searched for something, and switched to that site, then the next time you click to switch site, the previous search string is still in the box.

I'll look later today to address it, but I don't think it's really blocking all this.

**Comment by lsinger** [comment]
Since the IE11 / IE Edge compatibility issue is the only thing keeping this from being merged and both @marekhrabe and I have been unable to reproduce, I'll give this a go. Squash + merge! :shipit: 

@marekhrabe can you open issues for anything from above that you think we should be addressing in the future?

**Comment by ryanboren** [comment]
Testing this in Chrome and Firefox. In Chrome, Esc clears the search field when searching in the sidebar. In Firefox, Esc does not clear the search field.

**Comment by folletto** [comment]
@ryanboren As far as I know that's a difference in browser implementation for the `input type="search"` control. Firefox doesn't clear on ESC, Safari/Chrome do. 

Can you double-check it from [this demo](https://css-tricks.com/examples/SearchInputs/)?

**Comment by marekhrabe** [comment]
I can confirm that latest Firefox does *not* clear the search input on the demo page @folletto shared. I have same results with WordPress.com + Firefox = ESC does not clear the input.

There is an issue for that #5771 that I think we should reopen, as it still describes a present problem that was not fixed.
